pkg/graveler/graveler.go

```diff
@@ -871,13 +879,32 @@ func (g *Graveler) Get(ctx context.Context, repositoryID RepositoryID, ref Ref,
 	return g.CommittedManager.Get(ctx, repo.StorageNamespace, commit.MetaRangeID, key)
 }
 
-func (g *Graveler) Set(ctx context.Context, repositoryID RepositoryID, branchID BranchID, key Key, value Value) error {
+func (g *Graveler) Set(ctx context.Context, repositoryID RepositoryID, branchID BranchID, key Key, value Value, writeConditions ...WriteCondition) error {
 	_, err := g.branchLocker.Writer(ctx, repositoryID, branchID, func() (interface{}, error) {
 		branch, err := g.GetBranch(ctx, repositoryID, branchID)
 		if err != nil {
 			return nil, err
 		}
-		err = g.StagingManager.Set(ctx, branch.StagingToken, key, &value)
+		allowOverwrite := true
+		for _, cond := range writeConditions {
+			if cond.ifAbsent {
+				allowOverwrite = false
+			}
+		}
+		if !allowOverwrite {
```

Comment:
This is optimization as the ref manager will be passed override=false and it should fail in this case.